### PR TITLE
test: add comprehensive unit tests for TES and WES converters

### PR DESCRIPTION
**Pull request by Karanjot786**
This PR adds unit tests for the TES and WES converters and includes dummy data for testing purposes. The objective is to ensure that all components are working correctly and to provide a foundation for further testing and validation.

# Details

### Unit Tests for TES Converter

**Tests the conversion from TES to WRROC**

- Tests with complete data
- Tests with missing fields

**Tests the conversion from WRROC to TES**

- Tests with complete data
- Tests with missing fields

### Unit Tests for WES Converter

**Tests the conversion from WES to WRROC**

- Tests with complete data
- Tests with missing fields

**Tests the conversion from WRROC to WES**

- Tests with complete data
- Tests with missing fields

## Instructions for Running the Tests

**1. Install Dependencies:**
Ensure that all dependencies are installed:
```bash
poetry install
```

**2. Run the Unit Tests:**
Execute the following command to run the unit tests:
```bash
poetry run pytest tests/unit
```

## Run the TES to WRROC Conversion

Use the CLI to convert the TES input data to WRROC format.
```bash
poetry run python -m crategen.cli --input tests/data/input/tes_example.json --output tests/data/output/tes_to_wrroc_output.json --conversion-type tes-to-wrroc
```
Verify the output in `tes_to_wrroc_output.json`.

## Run the WES to WRROC Conversion
Use the CLI to convert the WES input data to WRROC format.
```bash
poetry run python -m crategen.cli --input tests/data/input/wes_example.json --output tests/data/output/wes_to_wrroc_output.json --conversion-type wes-to-wrroc
```
Verify the output in `wes_to_wrroc_output.json`.

### CODE REVIEW

File: tests/data/input/tes_example.json
Comment (uniqueg) [inline tests/data/input/tes_example.json:1]: I think we could easily make this more useful by picking values from a real toy TES run, i.e., put a container image that is actually available, put in a command that actually can be executed in that container, put input files that are permanently available on the web via HTTP(S) (e.g., the raw `README.md` and `LICENSE` files from this repo). Please open another issue for that.

It's a bit more tricky for the output, we can think about that in that issue.

File: tests/data/input/tes_example_with_missing_fields.json
Comment (uniqueg) [inline tests/data/input/tes_example_with_missing_fields.json:1]: Same as above. Also, make sure that all required params are provided (also in the tes example above). For example, the `command` param in `executors` is required (see https://github.com/ga4gh/task-execution-schemas/blob/develop/openapi/task_execution_service.openapi.yaml).

Also, I would name these files differently:
- `tes_minimal`
- `tes_full` (but then you'd also want to add resources etc.

File: tests/data/input/wes_example.json
Comment (uniqueg) [inline tests/data/input/wes_example.json:1]: It's probably also a good idea to have a minimal (required only) and full example (all params available).

File: tests/data/output/tes_to_wrroc_output.json
Comment (uniqueg) [inline tests/data/output/tes_to_wrroc_output.json:1]: Rename files according to suggested example names, e.g.:

- `wrroc_from_tes_minimal.json`
- `wrroc_from_tes_full.json`
- `wrroc_from_wes_minimal.json`
- `wrroc_from_wes_full.json`

**Review by uniqueg — CHANGES_REQUESTED** [review]
I gave it a quick look. Some suggestions. Also the Sourcery suggestions are really good, so please address them before re-review.

@SalihuDickson: Perhaps you want to also have a closer look at the unit test?

**Comment by Karanjot786** [comment]
Hi @uniqueg,

I have made the following updates based on your feedback:

## Updated Converters:

**WESConverter:**

- I excluded keys with None values from the conversion output.

**TESConverter:**

- Adjusted the conversion logic to ensure all required parameters are handled correctly.

**Enhanced Test Coverage**:
        

1. Added comprehensive test cases for both TES and WES converters, including:

- Edge cases for invalid data.
- Assertions for specific error messages.
- Tests for partially missing fields.
- Tests for edge cases with nested structures.
- Tests for invalid nested structures.

2. Updated test data to reflect real-world examples, including minimal and full examples for TES and WE

Please review the changes and let me know if any further adjustments are needed.

**Review by uniqueg — CHANGES_REQUESTED** [review]
I'm confused about the naming of these files still:
- Firstly, I am not sure about the distinction between inputs and outputs, especially given that there is not yet a clear definition of what inputs and outputs of each functionality should look like; without that and for now it seems to me that both WRROC and WES/TES payload files could be inputs _AND_ outputs - depending on the direction of the conversion, no?
- Probably it would make sense to indicate which WRROC profile a WRROC adheres to; we could use `ProcRC` for Process Run Crate `WfRC` for Workflow Run Crate, and `ProvRC` for Provenance Run Crate
- I think the `wrroc_from_` files should be WRROC files - not WES/TES payloads
- To be clear, I think we would need the following files:
  - `tes_minimal.json`: minimal `GET /tasks/{task_id}` response; input for TES > ProcRC conversion; output of WRROC > TES conversion
  - `tes_minimal.json`: full `GET /tasks/{task_id}` response; input for TES > ProcRC conversion; output of WRROC > TES conversion
  - `wes_minimal.json`: minimal WES `GET /runs/{id}` response; input for WES > WfRC conversion; output of WRROC > WES conversion
  - `wes_minimal.json`: full `GET /runs/{id}` response; input for WES > WfRC conversion; output of WRROC > wES conversion
  - `procrc_from_tes_minimal.json`: Process Run Crate packaged from minimal TES info; input for WRROC > TES conversion; output of TES > ProcRC conversion
  - `procrc_from_tes_minimal.json`: Process Run Crate packaged from full TES info; input for WRROC > TES conversion; output of TES > ProcRC conversion
  - `wfrc_from_wes_minimal.json`: Workflow Run Crate packaged from minimal WES info; input for WRROC > WES conversion; output of WES > WfRC conversion
  - `wfrc_from_wes_minimal.json`: Workflow Run Crate packaged from full WES info; input for WRROC > WES conversion; output of WES > WfRC conversion

File: tests/unit/test_tes_converter.py
Comment (SalihuDickson) [inline tests/unit/test_tes_converter.py:1]: you can group tests together using comments, that will help improve readability and maintainability further down the line.  

Also the way you're currently writing data validation tests will make it difficult to test for complex validation issues, implementing a pydantic model will reduce the need for simple validation tests like this, that way you can focus on possible, edge cases and you can create more complex instances of invalid data and test them at the same time.

File: tests/unit/test_wes_converter.py
Comment (SalihuDickson) [inline tests/unit/test_wes_converter.py:1]: you can group tests together using comments, that will help improve readability and maintainability further down the line.

Also the way you're currently writing data validation tests will make it difficult to test for complex validation issues, implementing a pydantic model will reduce the need for simple validation tests like this, that way you can focus on possible, edge cases and you can create more complex instances of invalid data and test them at the same time.

**Review by SalihuDickson — CHANGES_REQUESTED** [review]

File: tests/unit/test_tes_converter.py
Comment (uniqueg) [reply tests/unit/test_tes_converter.py:1]: +1

Small note: When using `pytest`, you can also group tests by using classes: https://docs.pytest.org/en/7.1.x/getting-started.html#group-multiple-tests-in-a-class

**Comment by Karanjot786** [comment]
This PR includes updates based on the mentor's feedback to improve data validation and readability:

**Updated Converters with Pydantic Models:**
        I used Pydantic models for data validation in both TES and WES converters.
        It simplified validation logic by leveraging Pydantic's built-in validation mechanisms.

**Updated Test Cases:**
        Grouped related test cases with comments to enhance readability and maintainability.
        Refactored data validation tests to utilize Pydantic models, reducing redundancy and focusing on edge cases.

**Renamed Test Files:**
        Ensured consistency in naming conventions as per mentor feedback.
        Updated the following files:
            `tes_minimal.json`
           `tes_full.json`
            `wes_minimal.json`
            `wes_full.json`
            `procrc_from_tes_minimal.json`
            `procrc_from_tes_full.json`
            `wfrc_from_wes_minimal.json`
            `wfrc_from_wes_full.json`

File: crategen/models.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    id: str
+    name: str
+    description: Optional[str] = ""
+    instrument: Optional[str] = None
+    object: List[WRROCInputs]
+    result: List[WRROCOutputs]
+    startTime: Optional[str] = None
+    endTime: Optional[str] = None
+
+    @validator('id')
```
Comment (SalihuDickson) [inline crategen/models.py:70]: these validators are redundant and cause unnecessary overhead, pydantic will perform automatic validation when you set a type for the fields, which you have already done above.

File: crategen/models.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+class WRROCDataWES(BaseModel):
+    id: str
+    name: str
+    status: str
+    result: List[WRROCOutputs]
+    startTime: Optional[str] = None
+    endTime: Optional[str] = None
+
+    @root_validator(pre=True)
```
Comment (SalihuDickson) [inline crategen/models.py:90]: The comment for the WES Data Model applies here as well

**Review by SalihuDickson — CHANGES_REQUESTED** [review]
took a quick look at the models and found somethings I'd like you to take a look at

File: crategen/models.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    location: str
+    name: str
+
+class WESData(BaseModel):
+    run_id: str
+    run_log: WESRunLog
+    state: str
+    outputs: List[WESOutputs]
+
+    @root_validator(pre=True)
```
[the remark this answers is not in the document]
Comment (Karanjot786) [reply crategen/models.py:44]: Thank you for the feedback. I've added extra = "forbid" to the models for better readability and error handling. I'll ensure all required fields for WES Data are included to prevent errors when running the CLI with wes_full.json.

File: crategen/models.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
-from pydantic import BaseModel, Field, validator,root_validator
-from typing import List, Optional, Dict, Any
+from pydantic import BaseModel, Field, validator, root_validator
+from typing import List, Optional, Dict
```
Comment (uniqueg) [inline crategen/models.py:2]: When using Python 3.9 or newer (which I hope you do), please use `list` and `dict` instead of the uppercase variants `List` and `Dict` for type hints and Pydantic models.

Comment (uniqueg) [reply crategen/models.py:2]: See https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections

**Review by uniqueg — CHANGES_REQUESTED** [review]

**Comment by uniqueg** [comment]
Please note also that this PR does now way more than adding unit tests. Please consider splitting this up into two PRs - the first for the code and model changes, the second one for adding unit tests.

File: crategen/converter_manager.py [hunk 5]
```diff
@@ -11,3 +11,9 @@ def convert_tes_to_wrroc(self, tes_data):
 
     def convert_wes_to_wrroc(self, wes_data):
         return self.wes_converter.convert_to_wrroc(wes_data)
+
+    def convert_wrroc_to_tes(self, wrroc_data):
+        return self.tes_converter.convert_from_wrroc(wrroc_data)
+
+    def convert_wrroc_to_wes(self, wrroc_data):
+        return self.wes_converter.convert_from_wrroc(wrroc_data)
```
[the remark this answers is not in the document]
Comment (Karanjot786) [reply crategen/converter_manager.py:19]: Thank you for the feedback. I will rename the arguments to data in all cases for consistency. I will also ensure that all methods/functions include type hints and Google-style docstrings in the next PR as per the contributor guide. I understand the importance of well-documented and typed code.

File: crategen/converters/tes_converter.py [hunk 6]
```diff
@@ -25,7 +25,9 @@ def convert_to_wrroc(self, tes_data):
 
     def convert_from_wrroc(self, wrroc_data):
         try:
-            validated_wrroc_data = WRROCData(**wrroc_data)
+            # Filter only the fields relevant to WRROCData
+            wrroc_filtered_data = {key: wrroc_data.get(key) for key in WRROCData.__fields__ if key in wrroc_data}
+            validated_wrroc_data = WRROCData(**wrroc_filtered_data)
         except ValidationError as e:
             raise ValueError(f"Invalid WRROC data: {e}")
```
[the remark this answers is not in the document]
Comment (Karanjot786) [reply crategen/converters/tes_converter.py:32]: I will address exception chaining in all relevant cases in a future PR to follow best practices as you suggested.

File: crategen/converters/tes_converter.py [hunk 7]
```diff
@@ -25,7 +25,9 @@ def convert_to_wrroc(self, tes_data):
 
     def convert_from_wrroc(self, wrroc_data):
         try:
-            validated_wrroc_data = WRROCData(**wrroc_data)
+            # Filter only the fields relevant to WRROCData
```
[the remark this answers is not in the document]
Comment (Karanjot786) [reply crategen/converters/tes_converter.py:28]: I will remove the redundant comment in the next PR to improve code readability.

File: crategen/converters/tes_converter.py [hunk 8]
```diff
@@ -25,7 +25,9 @@ def convert_to_wrroc(self, tes_data):
...
+            wrroc_filtered_data = {key: wrroc_data.get(key) for key in WRROCData.__fields__ if key in wrroc_data}
```
[the remark this answers is not in the document]
Comment (Karanjot786) [reply crategen/converters/tes_converter.py:29]: I will align the behavior of TES and WES converters to ensure consistent handling of irrelevant/extra fields by filtering them out. Additionally, I will standardize the use of `WRROCData` and `WRROCDataWES` naming conventions.

File: crategen/converters/wes_converter.py [hunk 9]
```diff
@@ -23,8 +23,15 @@ def convert_to_wrroc(self, wes_data):
...
+
+        if unexpected_fields:
+            raise ValueError(f"Unexpected fields in WRROC data: {unexpected_fields}")
+
         try:
-            wrroc_model = WRROCDataWES(**wrroc_data)
+            wrroc_filtered_data = {key: wrroc_data.get(key) for key in WRROCDataWES.__fields__}
+            wrroc_model = WRROCDataWES(**wrroc_filtered_data)
         except ValidationError as e:
             raise ValueError(f"Invalid WRROC data: {e}")
```
[the remark this answers is not in the document]
Comment (Karanjot786) [reply crategen/converters/wes_converter.py:36]: I will address these comments in the next PR, applying the same changes to both TES and WES converters for consistency.

File: crategen/models.py
[the remark this answers is not in the document]
Comment (Karanjot786) [reply crategen/models.py:1]: I will implement an external WRROC validator function to ensure WRROC data validity and avoid code repetition. Additionally, I will define separate models for WRROCDataTES and WRROCDataWES for specific validation in the respective converters.

File: crategen/models.py (hunk 4 above)
Comment (Karanjot786) [reply crategen/models.py:2]: I will update the type hints to use list and dict instead of List and Dict to align with Python 3.9+ standards.

**Comment by Karanjot786** [comment]
Hi @uniqueg I acknowledge the PR scope creep. I will split changes into smaller, more focused PRs to streamline the review process. Thank you for the guidance.

**Comment by Karanjot786** [comment]
Hi @uniqueg and @SalihuDickson,

I have addressed the feedback provided in the latest review. Here are the changes made in this PR:


1. **Argument Naming Consistency**:
    - Renamed the arguments `tes_data`, `wes_data`, and `wrroc_data` to `data` in all function definitions within the converters and manager to make the code more expressive and consistent.

2. **Type Hints and Google-Style Docstrings**:
    - Added type hints for all method arguments and return types across the codebase, except for test cases.
    - Added Google-style docstrings for all methods, classes, modules, and packages, including test cases. Refer to the contributor guide for details.

3. **Exception Chaining**:
    - Implemented exception chaining throughout the codebase for better error traceability. Replaced current exception handling with the recommended practice.

### Changes in Next PR:

1. **Type Hints and Google-Style Docstrings**:
    - Continuing with adding and refining type hints and Google-style docstrings as outlined in the contributor guide.

2. **Exception Chaining**:
    - Further refine and ensure all exceptions are handled with exception chaining for enhanced traceability.

I will continue refining type hints and docstrings and ensuring all exceptions are handled with exception chaining in the next PR.

Thank you for your guidance!

Best regards,
Karanjot Singh

File: crategen/converters/tes_converter.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (uniqueg) [reply crategen/converters/tes_converter.py:31]: Okay. Please create an issue for that and then close this conversation.

File: crategen/models.py
Comment (uniqueg) [inline crategen/models.py:1]: Please double check your `WRROCData`, `WRROCDataTES` and `WRROCDataWES` models according to the mapping you did at the beginning of the project.

- We want to use **`WRROCData`** to validate _any_ WRROC entity. However, given that there are three different profiles (Process Run, Workflow Run, Provenance Run), we want to have three different models. Luckily, given that these profiles are hierarchical (with Process Run being the least and Provenance Run being the most complex), we can make use of inheritance to avoid repeating ourselves. So you can start with the least complex one (`WRROCProcess`) and include all of its defined properties. None of its optional and all of its required properties should be required in your model. Additional properties should be allowed (see comment further below). Then go on with the next (`WRROCWorkflow`) and inherit from `WRROCProcess`. Only add additional properties defined for the Workflow Run profile that are not defined for Process Run (because the ones defined there will be inherited). If there are properties that are required for Workflow Run that were only optional for Process Run, override them. Finally, define `WRROCProvenance` in the same way, this time inheriting from `WRROCWorkflow`. Having these models, you can then write a validator that not only checks if a given object is a WRROC entity, but it will also tell you which profile(s) it adheres to. You can do so by first validating against `WRROCProvenance`. If it passes, you know that (1) the object is a WRROC entity and (2) it adheres to all profiles. If it does not pass, you validate against `WRROCWorkflow`. If that passes, you know that (1) the object is a WRROC entity and (2) it adheres to the Workflow Run and Process Run profiles. If it does not pass, you validate against `WRROCProcess`. If that passes, you know that (1) the object is a WRROC entity and (2) it adheres only to the Process Run profile. It it does not pass, you know that the object is not a WRROC entity. Having these models will be very useful to validate the input and also to automatically detect which conversions are possible.
- **`WRROCDataWES`** and **`WRROCDataTES`** should define ONLY those WRROC properties that are required to extract a WES and TES request from, respectively. We will use this to check whether we can proceed with the WRROC to WES conversion. Additional properties should be allowed (see comment further below). Given that both `WRROCDataWES` and `WRROCDataTES` are strict subsets of the core `WRROC` models above with respect to their properties (that is, e.g., all of the properties of `WRROCDataWES` will be defined somewhere in either of the `WRROC` profile models), you _could_ try to avoid repetitions by using one of the approaches discussed [here](https://github.com/pydantic/pydantic/issues/1010). However, if you do that, consider that some of the properties could be _required_ for `WRROCDataWES` and `WRROCDataTES`, but are only _optional_ in the WRROC profiles.

I think the models and the corresponding validators (including unit tests etc.) should be a single PR.

File: crategen/validators.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    Returns:
+        WRROCData: The validated WRROC data.
+
+    Raises:
+        ValueError: If the data is not valid WRROC data.
+    """
+    try:
+        return WRROCData(**data)
+    except ValidationError as e:
+        raise ValueError(f"Invalid WRROC data: {e}")
```
[the remark this answers is not in the document]
Comment (uniqueg) [reply crategen/validators.py:20]: Please work on the models first (see comments below and above), then rewrite your validators to account for what is written in those comments (especially the one below). But as I said, please do so in a clean, fresh PR that only changes/adds the 3 mentioned files. Make sure the code adheres to all good standards (docstrings, typing, error chaining etc), but only for the new code - the old code will be changed later.

**Comment by Karanjot786** [comment]
Hi @uniqueg and @SalihuDickson ,

Thank you for your detailed feedback. Here’s the plan to address the current comments and ensure a smooth review process:

1. **New Models and Validators**:
    - Create new models as described in your comments, complete with Google-style docstrings.
    - Implement corresponding validators in a dedicated validators module, including docstrings and type hints.
    - Develop comprehensive unit tests for the new validators.

I will open a new PR specifically for these tasks, ensuring they do not affect the current codebase, making it easy to review and merge.

After this initial PR is merged, we will proceed with the following steps:

1. Add docstrings to the rest of the codebase.
2. Add type hints to the rest of the codebase.
3. Connect the new models and validators with the existing code.
4. Add unit tests for all code.
5. Add integration tests.

Thank you for your guidance. I am looking forward to your further feedback.

Best regards,
Karanjot Singh
